cmd/vsphere-xcopy-volume-populator/internal/powerstore/powerstore.go

```diff
@@ -0,0 +1,297 @@
+					return populator.MappingContext{
+						"hostID":         host.ID,
+						"hostName":       host.Name,
+						"initiatorGroup": initiatorGroup,
+					}, nil
+				}
+			}
+		}
+	}
+
+	host, err := p.Client.GetHostByName(ctx, initiatorGroup)
+	if err != nil {
+		klog.Infof("Host %s not found, creating new host", initiatorGroup)
+		osType := gopowerstore.OSTypeEnumESXi
+		inits := make([]gopowerstore.InitiatorCreateModify, 0, len(adapterIds))
+		for _, a := range adapterIds {
+			pt := detectPortType(a)
+			inits = append(inits, gopowerstore.InitiatorCreateModify{
+				PortName: &a,
+				PortType: &pt,
+			})
+		}
+
+		createParams := &gopowerstore.HostCreate{
+			Name:       &initiatorGroup,
+			OsType:     &osType,
+			Initiators: &inits,
+		}
+
+		createResp, err := p.Client.CreateHost(ctx, createParams)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create host %s: %w", initiatorGroup, err)
+		}
+
+		// Get the created host
+		host, err = p.Client.GetHost(ctx, createResp.ID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get created host %s: %w", createResp.ID, err)
+		}
+
+		klog.Infof("Successfully created host %s with ID %s", initiatorGroup, host.ID)
+	} else {
+		klog.Infof("Found existing host %s with ID %s", initiatorGroup, host.ID)
+	}
+
+	// Step 2: Add initiators (adapter IDs) to the host
+	atLeastOneAdded := false
+
+	for _, adapterId := range adapterIds {
+		portType := detectPortType(adapterId)
+		klog.Infof("Processing adapter %s with type %s", adapterId, portType)
+
+		// Check if initiator already exists on this host
+		found := false
+		for _, initiator := range host.Initiators {
+			if initiator.PortName == adapterId {
+				klog.Infof("Initiator %s already exists on host %s", adapterId, initiatorGroup)
+				found = true
+				atLeastOneAdded = true
+				break
+			}
+		}
+
+		if !found {
+			// Add initiator to host
+			modifyParams := &gopowerstore.HostModify{
+				AddInitiators: &[]gopowerstore.InitiatorCreateModify{
+					{
+						PortName: &adapterId,
+						PortType: &portType,
+					},
+				},
+			}
+
+			_, err = p.Client.ModifyHost(ctx, modifyParams, host.ID)
+			if err != nil {
+				klog.Warningf("Failed to add initiator %s to host %s: %s", adapterId, initiatorGroup, err)
+				continue
+			}
+
+			klog.Infof("Successfully added %s initiator %s to host %s", portType, adapterId, initiatorGroup)
+			atLeastOneAdded = true
+		}
+	}
+
+	if !atLeastOneAdded {
+		return nil, fmt.Errorf("failed to add any adapters to host %s", initiatorGroup)
+	}
+
+	mappingContext["hostID"] = host.ID
+	mappingContext["hostName"] = host.Name
+	mappingContext["initiatorGroup"] = initiatorGroup
+
+	klog.Infof("Successfully ensured host %s with %d adapters", initiatorGroup, len(adapterIds))
+	return mappingContext, nil
+}
+
+// Helper function to detect port type from adapter ID
+func detectPortType(adapterId string) gopowerstore.InitiatorProtocolTypeEnum {
+	adapterId = strings.ToLower(strings.TrimSpace(adapterId))
```

Comment:
if the list of adapter IDs is not sanitized when queried from the ESX then please check the remote_esx.go and lets make sure it is sanitized there. I don't want implementation to have to go through this